bridge/server/services/data-service.ts

```diff
@@ -238,31 +238,39 @@ export class DataService {
     source?: KeptnService,
     resultType?: ResultTypes
   ): Promise<Trace[]> {
-    let traces: Trace[] = [];
-
-    for (let i = 0; i < services.length; i += this.MAX_PAGE_SIZE) {
-      const eventIds: string[] = [];
-      const maxLength = Math.min(i + this.MAX_PAGE_SIZE, services.length);
-      for (let y = i; y < maxLength; ++y) {
-        const latestServiceEvent = getServiceEvent(services[y]);
-        if (latestServiceEvent) {
-          // string concatenation is expensive; that's why we use an array here
-          eventIds.push(latestServiceEvent.eventId);
-        }
+    const traces: Trace[] = [];
+    let eventIds: string[] = [];
+    const fetchEvents = async (): Promise<void> => {
+      const response = await this.apiService.getTracesOfMultipleServices(
+        accessToken,
+        projectName,
+        eventType,
+        eventIds.join(',')
+      );
+      const events = response.data.events;
+      if (resultType || source) {
+        await this.checkAndSetEventsWithResult(accessToken, events, resultType, source);
       }
-      if (eventIds.length) {
-        const response = await this.apiService.getTracesOfMultipleServices(
-          accessToken,
-          projectName,
-          eventType,
-          eventIds.join(',')
-        );
-        if (resultType || source) {
-          await this.checkAndSetEventsWithResult(accessToken, response.data.events, resultType, source);
-        }
-        traces = [...traces, ...response.data.events];
+      traces.push(...events);
+      eventIds = [];
+    };
+
+    for (const service of services) {
+      const latestServiceEvent = getServiceEvent(service);
+      if (latestServiceEvent) {
+        // string concatenation is expensive; that's why we use an array here
+        eventIds.push(latestServiceEvent.eventId);
+      }
+
+      if (eventIds.length === this.MAX_PAGE_SIZE) {
+        await fetchEvents();
       }
     }
+    // get remaining events (mod 100)
+    if (eventIds.length) {
+      await fetchEvents();
+    }
+
     return traces;
```

Comment:
FYI the await inside the reduce will not work. The return value is `Promise<Trace[]>` and not Trace[]. `reduce` does not support any asynchronous tasks.

Reply:
Ok did not know. But still you can refactor the reducer and use the `chunk` function of lodash.

Reply:
Do we really want to add another dependency here? We currently don't use `lodash` for the bridge-server, only for the bridge-client.
It would also be easy to implement it in our own way
```ts
const traces: Trace[] = []
for (let i = 0; i < eventIds.length; i += this.MAX_PAGE_SIZE) {
  const eventIdChunk = eventIds.slice(i, i + this.MAX_PAGE_SIZE);
  const events = await fetchEvents(eventIdChunk);
  traces.push(...events);
}
```

Reply:
Yeah is also possible. The lodash function should work with every edge cases. Does yours?

Reply:
Alternatively you can collect the promises and use `await Promise.all(fetchEventsPromises)` to have them in an array.

Reply:
> Yeah is also possible. The lodash function should work with every edge cases. Does yours?

There aren't many cases so yes. Default case works and the other case where the second parameter `i+chunkSize` is too big for the remaining array also works

Reply:
> Alternatively you can collect the promises and use `await Promise.all(fetchEventsPromises)` to have them in an array. This would do the requests in parallel and improves performance.

Depends if the API calls should be one at a time or all at once. If all at once is possible (no rate limit or whatever). Then `Promise.all()` would be a performance boost...